Implement DH functionality

This is one of the types of keys that can be loaded through PKey. Follow-up of #234.

Needed: key generation, getting key parameters, relevant crypto operations

I think I'm being blocked by this. Can't find any way to simply get a public key from a generated Diffie-Hellman instance. I think I need to use `PKey::from_dh` and then `public_key_to_der/pem`, but

```Rust
error[E0599]: no method named `public_key_to_der` found for type `openssl::pkey::PKey<openssl::pkey::Params>` in the current scope
   --> lib.rs:105:28
    |
105 |         let key_der = pkey.public_key_to_der()?;
    |                            ^^^^^^^^^^^^^^^^^
    |
    = note: the method `public_key_to_der` exists but the following trait bounds were not satisfied:
            `openssl::pkey::Params : openssl::pkey::HasPublic`
```

--- a/openssl-sys/src/dh.rs
+++ b/openssl-sys/src/dh.rs
@@ -25,9 +25,9 @@ extern "C" {
     #[cfg(ossl102)]
     pub fn DH_get_2048_256() -> *mut DH;
 
-    #[cfg(any(ossl110, libressl273))]
+    #[cfg(any(ossl110, libressl270))]
     pub fn DH_set0_pqg(dh: *mut DH, p: *mut BIGNUM, q: *mut BIGNUM, g: *mut BIGNUM) -> c_int;
-    #[cfg(any(ossl110, libressl273))]
+    #[cfg(any(ossl110, libressl270))]
     pub fn DH_get0_pqg(
         dh: *const DH,
         p: *mut *const BIGNUM,
@@ -35,7 +35,6 @@ extern "C" {
         g: *mut *const BIGNUM,
     );
 
-    #[cfg(ossl102)]
+    #[cfg(any(ossl110, libressl270))]
     pub fn DH_get0_key(dh: *const DH, pub_key: *mut *const BIGNUM, priv_key: *mut *const BIGNUM);
-
 }

--- a/openssl-sys/src/dh.rs
+++ b/openssl-sys/src/dh.rs
@@ -4,6 +4,17 @@ extern "C" {
     pub fn DH_new() -> *mut DH;
     pub fn DH_free(dh: *mut DH);
 
+    pub fn DH_generate_parameters(
+        prime_len: c_int,
+        generator: c_int,
+        callback: Option<extern "C" fn(c_int, c_int, *mut c_void)>,
+        cb_arg: *mut c_void,
+    ) -> *mut DH;
+
+    pub fn DH_generate_key(dh: *mut DH) -> c_int;
+    pub fn DH_compute_key(key: *mut c_uchar, pub_key: *const BIGNUM, dh: *mut DH) -> c_int;
+    pub fn DH_size(dh: *const DH) -> c_int;
+
     pub fn d2i_DHparams(k: *mut *mut DH, pp: *mut *const c_uchar, length: c_long) -> *mut DH;
     pub fn i2d_DHparams(dh: *const DH, pp: *mut *mut c_uchar) -> c_int;
 
@@ -16,4 +27,7 @@ extern "C" {
 
     #[cfg(any(ossl110, libressl273))]
     pub fn DH_set0_pqg(dh: *mut DH, p: *mut BIGNUM, q: *mut BIGNUM, g: *mut BIGNUM) -> c_int;
+
+    #[cfg(ossl102)]
+    pub fn DH_get0_key(dh: *const DH, pub_key: *mut *const BIGNUM, priv_key: *mut *const BIGNUM);
 }

--- a/openssl-sys/src/dh.rs
+++ b/openssl-sys/src/dh.rs
@@ -27,7 +27,15 @@ extern "C" {
 
     #[cfg(any(ossl110, libressl273))]
     pub fn DH_set0_pqg(dh: *mut DH, p: *mut BIGNUM, q: *mut BIGNUM, g: *mut BIGNUM) -> c_int;
+    #[cfg(any(ossl110, libressl273))]
+    pub fn DH_get0_pqg(
+        dh: *const DH,
+        p: *mut *const BIGNUM,
+        q: *mut *const BIGNUM,
+        g: *mut *const BIGNUM,
+    );
 
     #[cfg(ossl102)]
     pub fn DH_get0_key(dh: *const DH, pub_key: *mut *const BIGNUM, priv_key: *mut *const BIGNUM);
+
 }

--- a/openssl/src/bn.rs
+++ b/openssl/src/bn.rs
@@ -114,7 +114,7 @@ foreign_type_and_impl_send_sync! {
     type CType = ffi::BIGNUM;
     fn drop = ffi::BN_free;
 
-    /// Dynamically sized large number impelementation
+    /// Dynamically sized large number implementation
     ///
     /// Perform large number mathematics.  Create a new BigNum
     /// with [`new`].  Perform standard mathematics on large numbers using

--- a/openssl/src/dh.rs
+++ b/openssl/src/dh.rs
@@ -8,28 +8,6 @@ use bn::{BigNum, BigNumRef};
 use pkey::{HasParams, HasPrivate, HasPublic, Params, Private};
 use {cvt, cvt_p};
 
-pub struct DhParams {
-    pub p: BigNum,
-    pub g: BigNum,
-    pub q: Option<BigNum>,
-}
-
-pub struct DhParamsRef<'a> {
-    pub p: &'a BigNumRef,
-    pub g: &'a BigNumRef,
-    pub q: Option<&'a BigNumRef>,
-}
-
-impl<'a> DhParamsRef<'a> {
-    pub fn to_owned(&self) -> Result<DhParams, ErrorStack> {
-        Ok(DhParams {
-            p: self.p.to_owned()?,
-            g: self.g.to_owned()?,
-            q: self.q.map(|q| q.to_owned()).transpose()?,
-        })
-    }
-}
-
 generic_foreign_type_and_impl_send_sync! {
     type CType = ffi::DH;
     fn drop = ffi::DH_free;
@@ -68,50 +46,73 @@ where
 
 impl Dh<Params> {
     pub fn from_params(p: BigNum, g: BigNum, q: BigNum) -> Result<Dh<Params>, ErrorStack> {
-        Self::from(DhParams { p, g, q: Some(q) })
+        Self::from_pqg(p, Some(q), g)
     }
 
-    /// Creates a DH instance based upon the given prime and generator params.
+    /// Creates a DH instance based upon the given primes and generator params.
     ///
     /// This corresponds to [`DH_new`] and [`DH_set0_pqg`].
     ///
     /// [`DH_new`]: https://www.openssl.org/docs/man1.1.0/crypto/DH_new.html
     /// [`DH_set0_pqg`]: https://www.openssl.org/docs/man1.1.0/crypto/DH_set0_pqg.html
-    pub fn from(params: DhParams) -> Result<Dh<Params>, ErrorStack> {
+    pub fn from_pqg(
+        prime_p: BigNum,
+        prime_q: Option<BigNum>,
+        generator: BigNum,
+    ) -> Result<Dh<Params>, ErrorStack> {
         unsafe {
             let dh = Dh::from_ptr(cvt_p(ffi::DH_new())?);
             cvt(DH_set0_pqg(
                 dh.0,
-                params.p.as_ptr(),
-                params.q.as_ref().map_or(ptr::null_mut(), |q| q.as_ptr()),
-                params.g.as_ptr(),
+                prime_p.as_ptr(),
+                prime_q.as_ref().map_or(ptr::null_mut(), |q| q.as_ptr()),
+                generator.as_ptr(),
             ))?;
-            mem::forget(params);
+            mem::forget((prime_p, prime_q, generator));
             Ok(dh)
         }
     }
 
-    /// Gets the prime and generator params from the DH instance.
+    /// Returns the prime `p` from the DH instance.
     ///
     /// This corresponds to [`DH_get0_pqg`].
     ///
     /// [`DH_get0_pqg`]: https://www.openssl.org/docs/man1.1.0/crypto/DH_get0_pqg.html
-    pub fn get_params(&self) -> DhParamsRef {
+    pub fn prime_p(&self) -> &BigNumRef {
         let mut p = ptr::null();
-        let mut g = ptr::null();
+        unsafe {
+            DH_get0_pqg(self.0, &mut p, ptr::null_mut(), ptr::null_mut());
+            BigNumRef::from_ptr(p as *mut _)
+        }
+    }
+
+    /// Returns the prime `q` from the DH instance.
+    ///
+    /// This corresponds to [`DH_get0_pqg`].
+    ///
+    /// [`DH_get0_pqg`]: https://www.openssl.org/docs/man1.1.0/crypto/DH_get0_pqg.html
+    pub fn prime_q(&self) -> Option<&BigNumRef> {
         let mut q = ptr::null();
         unsafe {
-            DH_get0_pqg(self.0, &mut p, &mut q, &mut g);
-            let q = if q.is_null() {
+            DH_get0_pqg(self.0, ptr::null_mut(), &mut q, ptr::null_mut());
+            if q.is_null() {
                 None
             } else {
                 Some(BigNumRef::from_ptr(q as *mut _))
-            };
-            DhParamsRef {
-                p: &BigNumRef::from_ptr(p as *mut _),
-                g: &BigNumRef::from_ptr(g as *mut _),
-                q,
             }
+        }
+    }
+
+    /// Returns the generator from the DH instance.
+    ///
+    /// This corresponds to [`DH_get0_pqg`].
+    ///
+    /// [`DH_get0_pqg`]: https://www.openssl.org/docs/man1.1.0/crypto/DH_get0_pqg.html
+    pub fn generator(&self) -> &BigNumRef {
+        let mut g = ptr::null();
+        unsafe {
+            DH_get0_pqg(self.0, ptr::null_mut(), ptr::null_mut(), &mut g);
+            BigNumRef::from_ptr(g as *mut _)
         }
     }
 
@@ -201,12 +202,12 @@ impl<T> Dh<T>
 where
     T: HasPublic,
 {
-    /// Gets the public key from the DH instance.
+    /// Returns the public key from the DH instance.
     ///
     /// This corresponds to [`DH_get0_key`].
     ///
     /// [`DH_get0_key`]: https://www.openssl.org/docs/man1.1.0/crypto/DH_get0_key.html
-    pub fn get_public_key(&self) -> &BigNumRef {
+    pub fn public_key(&self) -> &BigNumRef {
         let mut pub_key = ptr::null();
         unsafe {
             DH_get0_key(self.0, &mut pub_key, ptr::null_mut());
@@ -262,9 +263,15 @@ cfg_if! {
             q: *mut *const ffi::BIGNUM,
             g: *mut *const ffi::BIGNUM,
         ) {
-            *p = (*dh).p;
-            *q = (*dh).q;
-            *g = (*dh).g;
+            if !p.is_null() {
+                *p = (*dh).p;
+            }
+            if !q.is_null() {
+                *q = (*dh).q;
+            }
+            if !g.is_null() {
+                *g = (*dh).g;
+            }
         }
 
         #[allow(bad_style)]
@@ -302,7 +309,7 @@ mod tests {
     #[test]
     fn test_dh_params() {
         let mut ctx = SslContext::builder(SslMethod::tls()).unwrap();
-        let p = BigNum::from_hex_str(
+        let prime_p = BigNum::from_hex_str(
             "87A8E61DB4B6663CFFBBD19C651959998CEEF608660DD0F25D2CEED4435E3B00E00DF8F1D61957D4FAF7DF\
              4561B2AA3016C3D91134096FAA3BF4296D830E9A7C209E0C6497517ABD5A8A9D306BCF67ED91F9E6725B47\
              58C022E0B1EF4275BF7B6C5BFC11D45F9088B941F54EB1E59BB8BC39A0BF12307F5C4FDB70C581B23F76B6\
@@ -310,7 +317,7 @@ mod tests {
              140564251CCACB83E6B486F6B3CA3F7971506026C0B857F689962856DED4010ABD0BE621C3A3960A54E710\
              C375F26375D7014103A4B54330C198AF126116D2276E11715F693877FAD7EF09CADB094AE91E1A1597",
         ).unwrap();
-        let g = BigNum::from_hex_str(
+        let prime_q = BigNum::from_hex_str(
             "3FB32C9B73134D0B2E77506660EDBD484CA7B18F21EF205407F4793A1A0BA12510DBC15077BE463FFF4FED\
              4AAC0BB555BE3A6C1B0C6B47B1BC3773BF7E8C6F62901228F8C28CBB18A55AE31341000A650196F931C77A\
              57F2DDF463E5E9EC144B777DE62AAAB8A8628AC376D282D6ED3864E67982428EBC831D14348F6F2F9193B5\
@@ -318,23 +325,21 @@ mod tests {
              052588B9B7D2BBD2DF016199ECD06E1557CD0915B3353BBB64E0EC377FD028370DF92B52C7891428CDC67E\
              B6184B523D1DB246C32F63078490F00EF8D647D148D47954515E2327CFEF98C582664B4C0F6CC41659",
         ).unwrap();
-        let q = BigNum::from_hex_str(
+        let generator = BigNum::from_hex_str(
             "8CF83642A709A097B447997640129DA299B1A47D1EB3750BA308B0FE64F5FBD3",
         )
         .unwrap();
         let dh = Dh::from_params(
-            p.to_owned().unwrap(),
-            g.to_owned().unwrap(),
-            q.to_owned().unwrap(),
+            prime_p.to_owned().unwrap(),
+            generator.to_owned().unwrap(),
+            prime_q.to_owned().unwrap(),
         )
         .unwrap();
         ctx.set_tmp_dh(&dh).unwrap();
 
-        let params = dh.get_params();
-
-        assert_eq!(params.p, &p);
-        assert_eq!(params.g, &g);
-        assert_eq!(params.q.unwrap(), &q);
+        assert_eq!(dh.prime_p(), &prime_p);
+        assert_eq!(dh.prime_q().unwrap(), &prime_q);
+        assert_eq!(dh.generator(), &generator);
     }
 
     #[test]
@@ -359,23 +364,27 @@ mod tests {
         let dh1 = Dh::get_2048_224().unwrap().generate_key().unwrap();
         let dh2 = Dh::get_2048_224().unwrap().generate_key().unwrap();
 
-        let shared_a = dh1.compute_key(dh2.get_public_key()).unwrap();
-        let shared_b = dh2.compute_key(dh1.get_public_key()).unwrap();
+        let shared_a = dh1.compute_key(dh2.public_key()).unwrap();
+        let shared_b = dh2.compute_key(dh1.public_key()).unwrap();
 
         assert_eq!(shared_a, shared_b);
     }
 
     #[test]
     fn test_dh_generate_params_generate_key_compute_key() {
-        let dh_params1 = Dh::generate_params(1024, 2).unwrap();
-        let params = dh_params1.get_params().to_owned().unwrap();
-        let dh_params2 = Dh::from(params).unwrap();
+        let dh_params1 = Dh::generate_params(512, 2).unwrap();
+        let dh_params2 = Dh::from_pqg(
+            dh_params1.prime_p().to_owned().unwrap(),
+            None,
+            dh_params1.generator().to_owned().unwrap(),
+        )
+        .unwrap();
 
         let dh1 = dh_params1.generate_key().unwrap();
         let dh2 = dh_params2.generate_key().unwrap();
 
-        let shared_a = dh1.compute_key(dh2.get_public_key()).unwrap();
-        let shared_b = dh2.compute_key(dh1.get_public_key()).unwrap();
+        let shared_a = dh1.compute_key(dh2.public_key()).unwrap();
+        let shared_b = dh2.compute_key(dh1.public_key()).unwrap();
 
         assert_eq!(shared_a, shared_b);
     }

--- a/openssl/src/dh.rs
+++ b/openssl/src/dh.rs
@@ -194,7 +194,7 @@ mod tests {
     }
 
     #[test]
-    fn test_dh() {
+    fn test_dh_from_params() {
         let mut ctx = SslContext::builder(SslMethod::tls()).unwrap();
         let p = BigNum::from_hex_str(
             "87A8E61DB4B6663CFFBBD19C651959998CEEF608660DD0F25D2CEED4435E3B00E00DF8F1D61957D4FAF7DF\
@@ -234,5 +234,17 @@ mod tests {
         let dh = Dh::params_from_pem(params).unwrap();
         let der = dh.params_to_der().unwrap();
         Dh::params_from_der(&der).unwrap();
+    }
+
+    #[test]
+    fn test_dh_generate_compute_key() {
+        let dh1 = Dh::get_2048_224().unwrap().generate_key().unwrap();
+
+        let dh2 = Dh::get_2048_224().unwrap().generate_key().unwrap();
+
+        let shared_a = dh1.compute_key(dh2.get_public_key().unwrap()).unwrap();
+        let shared_b = dh2.compute_key(dh1.get_public_key().unwrap()).unwrap();
+
+        assert_eq!(shared_a, shared_b);
     }
 }

--- a/openssl/src/dh.rs
+++ b/openssl/src/dh.rs
@@ -8,6 +8,28 @@ use bn::{BigNum, BigNumRef};
 use pkey::{HasParams, HasPrivate, HasPublic, Params, Private};
 use {cvt, cvt_p};
 
+pub struct DhParams {
+    pub p: BigNum,
+    pub g: BigNum,
+    pub q: Option<BigNum>,
+}
+
+pub struct DhParamsRef<'a> {
+    pub p: &'a BigNumRef,
+    pub g: &'a BigNumRef,
+    pub q: Option<&'a BigNumRef>,
+}
+
+impl<'a> DhParamsRef<'a> {
+    pub fn to_owned(&self) -> Result<DhParams, ErrorStack> {
+        Ok(DhParams {
+            p: self.p.to_owned()?,
+            g: self.g.to_owned()?,
+            q: self.q.map(|q| q.to_owned()).transpose()?,
+        })
+    }
+}
+
 generic_foreign_type_and_impl_send_sync! {
     type CType = ffi::DH;
     fn drop = ffi::DH_free;
@@ -45,21 +67,36 @@ where
 }
 
 impl Dh<Params> {
-    pub fn from_params(p: BigNum, g: BigNum, q: Option<BigNum>) -> Result<Dh<Params>, ErrorStack> {
+    pub fn from_params(p: BigNum, g: BigNum, q: BigNum) -> Result<Dh<Params>, ErrorStack> {
+        Self::from(DhParams { p, g, q: Some(q) })
+    }
+
+    /// Creates a DH instance based upon the given prime and generator params.
+    ///
+    /// This corresponds to [`DH_new`] and [`DH_set0_pqg`].
+    ///
+    /// [`DH_new`]: https://www.openssl.org/docs/man1.1.0/crypto/DH_new.html
+    /// [`DH_set0_pqg`]: https://www.openssl.org/docs/man1.1.0/crypto/DH_set0_pqg.html
+    pub fn from(params: DhParams) -> Result<Dh<Params>, ErrorStack> {
         unsafe {
             let dh = Dh::from_ptr(cvt_p(ffi::DH_new())?);
             cvt(DH_set0_pqg(
                 dh.0,
-                p.as_ptr(),
-                q.as_ref().map_or(ptr::null_mut(), |q| q.as_ptr()),
-                g.as_ptr(),
+                params.p.as_ptr(),
+                params.q.as_ref().map_or(ptr::null_mut(), |q| q.as_ptr()),
+                params.g.as_ptr(),
             ))?;
-            mem::forget((p, g, q));
+            mem::forget(params);
             Ok(dh)
         }
     }
 
-    pub fn get_params(&self) -> (&BigNumRef, &BigNumRef, Option<&BigNumRef>) {
+    /// Gets the prime and generator params from the DH instance.
+    ///
+    /// This corresponds to [`DH_get0_pqg`].
+    ///
+    /// [`DH_get0_pqg`]: https://www.openssl.org/docs/man1.1.0/crypto/DH_get0_pqg.html
+    pub fn get_params(&self) -> DhParamsRef {
         let mut p = ptr::null();
         let mut g = ptr::null();
         let mut q = ptr::null();
@@ -70,11 +107,11 @@ impl Dh<Params> {
             } else {
                 Some(BigNumRef::from_ptr(q as *mut _))
             };
-            (
-                &BigNumRef::from_ptr(p as *mut _),
-                &BigNumRef::from_ptr(g as *mut _),
+            DhParamsRef {
+                p: &BigNumRef::from_ptr(p as *mut _),
+                g: &BigNumRef::from_ptr(g as *mut _),
                 q,
-            )
+            }
         }
     }
 
@@ -172,7 +209,7 @@ where
     pub fn get_public_key(&self) -> &BigNumRef {
         let mut pub_key = ptr::null();
         unsafe {
-            ffi::DH_get0_key(self.0, &mut pub_key, ptr::null_mut());
+            DH_get0_key(self.0, &mut pub_key, ptr::null_mut());
             BigNumRef::from_ptr(pub_key as *mut _)
         }
     }
@@ -203,7 +240,7 @@ where
 
 cfg_if! {
     if #[cfg(any(ossl110, libressl270))] {
-        use ffi::{DH_set0_pqg, DH_get0_pqg};
+        use ffi::{DH_set0_pqg, DH_get0_pqg, DH_get0_key};
     } else {
         #[allow(bad_style)]
         unsafe fn DH_set0_pqg(
@@ -228,6 +265,20 @@ cfg_if! {
             *p = (*dh).p;
             *q = (*dh).q;
             *g = (*dh).g;
+        }
+
+        #[allow(bad_style)]
+        unsafe fn DH_get0_key(
+            dh: *mut ffi::DH,
+            pub_key: *mut *const ffi::BIGNUM,
+            priv_key: *mut *const ffi::BIGNUM,
+        ) {
+            if !pub_key.is_null() {
+                *pub_key = (*dh).pub_key;
+            }
+            if !priv_key.is_null() {
+                *priv_key = (*dh).priv_key;
+            }
         }
     }
 }
@@ -274,16 +325,16 @@ mod tests {
         let dh = Dh::from_params(
             p.to_owned().unwrap(),
             g.to_owned().unwrap(),
-            Some(q.to_owned().unwrap()),
+            q.to_owned().unwrap(),
         )
         .unwrap();
         ctx.set_tmp_dh(&dh).unwrap();
 
-        let (p1, g1, q1) = dh.get_params();
+        let params = dh.get_params();
 
-        assert_eq!(p1, &p);
-        assert_eq!(g1, &g);
-        assert_eq!(q1.unwrap(), &q);
+        assert_eq!(params.p, &p);
+        assert_eq!(params.g, &g);
+        assert_eq!(params.q.unwrap(), &q);
     }
 
     #[test]
@@ -303,6 +354,7 @@ mod tests {
     }
 
     #[test]
+    #[cfg(ossl102)]
     fn test_dh_generate_key_compute_key() {
         let dh1 = Dh::get_2048_224().unwrap().generate_key().unwrap();
         let dh2 = Dh::get_2048_224().unwrap().generate_key().unwrap();
@@ -316,13 +368,8 @@ mod tests {
     #[test]
     fn test_dh_generate_params_generate_key_compute_key() {
         let dh_params1 = Dh::generate_params(1024, 2).unwrap();
-        let (prime, generator, _) = dh_params1.get_params();
-        let dh_params2 = Dh::from_params(
-            prime.to_owned().unwrap(),
-            generator.to_owned().unwrap(),
-            None,
-        )
-        .unwrap();
+        let params = dh_params1.get_params().to_owned().unwrap();
+        let dh_params2 = Dh::from(params).unwrap();
 
         let dh1 = dh_params1.generate_key().unwrap();
         let dh2 = dh_params2.generate_key().unwrap();

--- a/openssl/src/dh.rs
+++ b/openssl/src/dh.rs
@@ -45,12 +45,36 @@ where
 }
 
 impl Dh<Params> {
-    pub fn from_params(p: BigNum, g: BigNum, q: BigNum) -> Result<Dh<Params>, ErrorStack> {
+    pub fn from_params(p: BigNum, g: BigNum, q: Option<BigNum>) -> Result<Dh<Params>, ErrorStack> {
         unsafe {
             let dh = Dh::from_ptr(cvt_p(ffi::DH_new())?);
-            cvt(DH_set0_pqg(dh.0, p.as_ptr(), q.as_ptr(), g.as_ptr()))?;
+            cvt(DH_set0_pqg(
+                dh.0,
+                p.as_ptr(),
+                q.as_ref().map_or(ptr::null_mut(), |q| q.as_ptr()),
+                g.as_ptr(),
+            ))?;
             mem::forget((p, g, q));
             Ok(dh)
+        }
+    }
+
+    pub fn get_params(&self) -> Result<(BigNum, BigNum, Option<BigNum>), ErrorStack> {
+        let mut p = ptr::null();
+        let mut g = ptr::null();
+        let mut q = ptr::null();
+        unsafe {
+            ffi::DH_get0_pqg(self.0, &mut p, &mut q, &mut g);
+            let q = if q.is_null() {
+                None
+            } else {
+                Some(BigNum::from_ptr(cvt_p(ffi::BN_dup(q))?))
+            };
+            Ok((
+                BigNum::from_ptr(cvt_p(ffi::BN_dup(p))?),
+                BigNum::from_ptr(cvt_p(ffi::BN_dup(g))?),
+                q,
+            ))
         }
     }
 
@@ -132,9 +156,8 @@ where
 {
     pub fn get_public_key(&self) -> Result<BigNum, ErrorStack> {
         let mut pub_key = ptr::null();
-        let mut priv_key = ptr::null();
         unsafe {
-            ffi::DH_get0_key(self.0, &mut pub_key, &mut priv_key);
+            ffi::DH_get0_key(self.0, &mut pub_key, ptr::null_mut());
             Ok(BigNum::from_ptr(cvt_p(ffi::BN_dup(pub_key))?))
         }
     }
@@ -194,7 +217,7 @@ mod tests {
     }
 
     #[test]
-    fn test_dh_from_params() {
+    fn test_dh_params() {
         let mut ctx = SslContext::builder(SslMethod::tls()).unwrap();
         let p = BigNum::from_hex_str(
             "87A8E61DB4B6663CFFBBD19C651959998CEEF608660DD0F25D2CEED4435E3B00E00DF8F1D61957D4FAF7DF\
@@ -216,8 +239,19 @@ mod tests {
             "8CF83642A709A097B447997640129DA299B1A47D1EB3750BA308B0FE64F5FBD3",
         )
         .unwrap();
-        let dh = Dh::from_params(p, g, q).unwrap();
+        let dh = Dh::from_params(
+            p.to_owned().unwrap(),
+            g.to_owned().unwrap(),
+            Some(q.to_owned().unwrap()),
+        )
+        .unwrap();
         ctx.set_tmp_dh(&dh).unwrap();
+
+        let (p1, g1, q1) = dh.get_params().unwrap();
+
+        assert_eq!(p1, p);
+        assert_eq!(g1, g);
+        assert_eq!(q1.unwrap(), q);
     }
 
     #[test]
@@ -237,10 +271,26 @@ mod tests {
     }
 
     #[test]
-    fn test_dh_generate_compute_key() {
+    fn test_dh_generate_key_compute_key() {
         let dh1 = Dh::get_2048_224().unwrap().generate_key().unwrap();
 
         let dh2 = Dh::get_2048_224().unwrap().generate_key().unwrap();
+
+        let shared_a = dh1.compute_key(dh2.get_public_key().unwrap()).unwrap();
+        let shared_b = dh2.compute_key(dh1.get_public_key().unwrap()).unwrap();
+
+        assert_eq!(shared_a, shared_b);
+    }
+
+    #[test]
+    fn test_dh_generate_params_generate_key_compute_key() {
+        let dh_params1 = Dh::generate_params(1024, 2).unwrap();
+
+        let (prime, generator, _) = dh_params1.get_params().unwrap();
+        let dh1 = dh_params1.generate_key().unwrap();
+
+        let dh_params2 = Dh::from_params(prime, generator, None).unwrap();
+        let dh2 = dh_params2.generate_key().unwrap();
 
         let shared_a = dh1.compute_key(dh2.get_public_key().unwrap()).unwrap();
         let shared_b = dh2.compute_key(dh1.get_public_key().unwrap()).unwrap();

--- a/openssl/src/dh.rs
+++ b/openssl/src/dh.rs
@@ -4,7 +4,7 @@ use foreign_types::{ForeignType, ForeignTypeRef};
 use std::mem;
 use std::ptr;
 
-use bn::BigNum;
+use bn::{BigNum, BigNumRef};
 use pkey::{HasParams, HasPrivate, HasPublic, Params, Private};
 use {cvt, cvt_p};
 
@@ -59,25 +59,30 @@ impl Dh<Params> {
         }
     }
 
-    pub fn get_params(&self) -> Result<(BigNum, BigNum, Option<BigNum>), ErrorStack> {
+    pub fn get_params(&self) -> (&BigNumRef, &BigNumRef, Option<&BigNumRef>) {
         let mut p = ptr::null();
         let mut g = ptr::null();
         let mut q = ptr::null();
         unsafe {
-            ffi::DH_get0_pqg(self.0, &mut p, &mut q, &mut g);
+            DH_get0_pqg(self.0, &mut p, &mut q, &mut g);
             let q = if q.is_null() {
                 None
             } else {
-                Some(BigNum::from_ptr(cvt_p(ffi::BN_dup(q))?))
+                Some(BigNumRef::from_ptr(q as *mut _))
             };
-            Ok((
-                BigNum::from_ptr(cvt_p(ffi::BN_dup(p))?),
-                BigNum::from_ptr(cvt_p(ffi::BN_dup(g))?),
+            (
+                &BigNumRef::from_ptr(p as *mut _),
+                &BigNumRef::from_ptr(g as *mut _),
                 q,
-            ))
+            )
         }
     }
 
+    /// Generates DH params based on the given `prime_len` and a fixed `generator` value.
+    ///
+    /// This corresponds to [`DH_generate_parameters`].
+    ///
+    /// [`DH_generate_parameters`]: https://www.openssl.org/docs/man1.1.0/crypto/DH_generate_parameters.html
     pub fn generate_params(prime_len: u32, generator: u32) -> Result<Dh<Params>, ErrorStack> {
         unsafe {
             Ok(Dh::from_ptr(cvt_p(ffi::DH_generate_parameters(
@@ -89,6 +94,11 @@ impl Dh<Params> {
         }
     }
 
+    /// Generates a public and a private key based on the DH params.
+    ///
+    /// This corresponds to [`DH_generate_key`].
+    ///
+    /// [`DH_generate_key`]: https://www.openssl.org/docs/man1.1.0/crypto/DH_generate_key.html
     pub fn generate_key(self) -> Result<Dh<Private>, ErrorStack> {
         unsafe {
             let dh_ptr = self.0;
@@ -154,11 +164,16 @@ impl<T> Dh<T>
 where
     T: HasPublic,
 {
-    pub fn get_public_key(&self) -> Result<BigNum, ErrorStack> {
+    /// Gets the public key from the DH instance.
+    ///
+    /// This corresponds to [`DH_get0_key`].
+    ///
+    /// [`DH_get0_key`]: https://www.openssl.org/docs/man1.1.0/crypto/DH_get0_key.html
+    pub fn get_public_key(&self) -> &BigNumRef {
         let mut pub_key = ptr::null();
         unsafe {
             ffi::DH_get0_key(self.0, &mut pub_key, ptr::null_mut());
-            Ok(BigNum::from_ptr(cvt_p(ffi::BN_dup(pub_key))?))
+            BigNumRef::from_ptr(pub_key as *mut _)
         }
     }
 }
@@ -167,7 +182,12 @@ impl<T> Dh<T>
 where
     T: HasPrivate,
 {
-    pub fn compute_key(&self, public_key: BigNum) -> Result<Vec<u8>, ErrorStack> {
+    /// Computes a shared secret from the own private key and the given `public_key`.
+    ///
+    /// This corresponds to [`DH_compute_key`].
+    ///
+    /// [`DH_compute_key`]: https://www.openssl.org/docs/man1.1.0/crypto/DH_compute_key.html
+    pub fn compute_key(&self, public_key: &BigNumRef) -> Result<Vec<u8>, ErrorStack> {
         unsafe {
             let key_len = ffi::DH_size(self.0);
             let mut key = vec![0u8; key_len as usize];
@@ -182,8 +202,8 @@ where
 }
 
 cfg_if! {
-    if #[cfg(any(ossl110, libressl273))] {
-        use ffi::DH_set0_pqg;
+    if #[cfg(any(ossl110, libressl270))] {
+        use ffi::{DH_set0_pqg, DH_get0_pqg};
     } else {
         #[allow(bad_style)]
         unsafe fn DH_set0_pqg(
@@ -196,6 +216,18 @@ cfg_if! {
             (*dh).q = q;
             (*dh).g = g;
             1
+        }
+
+        #[allow(bad_style)]
+        unsafe fn DH_get0_pqg(
+            dh: *mut ffi::DH,
+            p: *mut *const ffi::BIGNUM,
+            q: *mut *const ffi::BIGNUM,
+            g: *mut *const ffi::BIGNUM,
+        ) {
+            *p = (*dh).p;
+            *q = (*dh).q;
+            *g = (*dh).g;
         }
     }
 }
@@ -247,11 +279,11 @@ mod tests {
         .unwrap();
         ctx.set_tmp_dh(&dh).unwrap();
 
-        let (p1, g1, q1) = dh.get_params().unwrap();
+        let (p1, g1, q1) = dh.get_params();
 
-        assert_eq!(p1, p);
-        assert_eq!(g1, g);
-        assert_eq!(q1.unwrap(), q);
+        assert_eq!(p1, &p);
+        assert_eq!(g1, &g);
+        assert_eq!(q1.unwrap(), &q);
     }
 
     #[test]
@@ -273,11 +305,10 @@ mod tests {
     #[test]
     fn test_dh_generate_key_compute_key() {
         let dh1 = Dh::get_2048_224().unwrap().generate_key().unwrap();
-
         let dh2 = Dh::get_2048_224().unwrap().generate_key().unwrap();
 
-        let shared_a = dh1.compute_key(dh2.get_public_key().unwrap()).unwrap();
-        let shared_b = dh2.compute_key(dh1.get_public_key().unwrap()).unwrap();
+        let shared_a = dh1.compute_key(dh2.get_public_key()).unwrap();
+        let shared_b = dh2.compute_key(dh1.get_public_key()).unwrap();
 
         assert_eq!(shared_a, shared_b);
     }
@@ -285,15 +316,19 @@ mod tests {
     #[test]
     fn test_dh_generate_params_generate_key_compute_key() {
         let dh_params1 = Dh::generate_params(1024, 2).unwrap();
+        let (prime, generator, _) = dh_params1.get_params();
+        let dh_params2 = Dh::from_params(
+            prime.to_owned().unwrap(),
+            generator.to_owned().unwrap(),
+            None,
+        )
+        .unwrap();
 
-        let (prime, generator, _) = dh_params1.get_params().unwrap();
         let dh1 = dh_params1.generate_key().unwrap();
-
-        let dh_params2 = Dh::from_params(prime, generator, None).unwrap();
         let dh2 = dh_params2.generate_key().unwrap();
 
-        let shared_a = dh1.compute_key(dh2.get_public_key().unwrap()).unwrap();
-        let shared_b = dh2.compute_key(dh1.get_public_key().unwrap()).unwrap();
+        let shared_a = dh1.compute_key(dh2.get_public_key()).unwrap();
+        let shared_b = dh2.compute_key(dh1.get_public_key()).unwrap();
 
         assert_eq!(shared_a, shared_b);
     }

--- a/openssl/src/dh.rs
+++ b/openssl/src/dh.rs
@@ -5,7 +5,7 @@ use std::mem;
 use std::ptr;
 
 use bn::BigNum;
-use pkey::{HasParams, Params};
+use pkey::{HasParams, HasPrivate, HasPublic, Params, Private};
 use {cvt, cvt_p};
 
 generic_foreign_type_and_impl_send_sync! {
@@ -51,6 +51,26 @@ impl Dh<Params> {
             cvt(DH_set0_pqg(dh.0, p.as_ptr(), q.as_ptr(), g.as_ptr()))?;
             mem::forget((p, g, q));
             Ok(dh)
+        }
+    }
+
+    pub fn generate_params(prime_len: u32, generator: u32) -> Result<Dh<Params>, ErrorStack> {
+        unsafe {
+            Ok(Dh::from_ptr(cvt_p(ffi::DH_generate_parameters(
+                prime_len as i32,
+                generator as i32,
+                None,
+                ptr::null_mut(),
+            ))?))
+        }
+    }
+
+    pub fn generate_key(self) -> Result<Dh<Private>, ErrorStack> {
+        unsafe {
+            let dh_ptr = self.0;
+            cvt(ffi::DH_generate_key(dh_ptr))?;
+            mem::forget(self);
+            Ok(Dh::from_ptr(dh_ptr))
         }
     }
 
@@ -102,6 +122,38 @@ impl Dh<Params> {
         unsafe {
             ffi::init();
             cvt_p(ffi::DH_get_2048_256()).map(|p| Dh::from_ptr(p))
+        }
+    }
+}
+
+impl<T> Dh<T>
+where
+    T: HasPublic,
+{
+    pub fn get_public_key(&self) -> Result<BigNum, ErrorStack> {
+        let mut pub_key = ptr::null();
+        let mut priv_key = ptr::null();
+        unsafe {
+            ffi::DH_get0_key(self.0, &mut pub_key, &mut priv_key);
+            Ok(BigNum::from_ptr(cvt_p(ffi::BN_dup(pub_key))?))
+        }
+    }
+}
+
+impl<T> Dh<T>
+where
+    T: HasPrivate,
+{
+    pub fn compute_key(&self, public_key: BigNum) -> Result<Vec<u8>, ErrorStack> {
+        unsafe {
+            let key_len = ffi::DH_size(self.0);
+            let mut key = vec![0u8; key_len as usize];
+            cvt(ffi::DH_compute_key(
+                key.as_mut_ptr(),
+                public_key.as_ptr(),
+                self.0,
+            ))?;
+            Ok(key)
         }
     }
 }

